cmd/lncli/profile.go

```diff
@@ -0,0 +1,324 @@
+				"macaroon: %v", err)
+		}
+	}
+
+	// Parse the macaroon data into its native struct.
+	mac := &macaroon.Macaroon{}
+	if err := mac.UnmarshalBinary(macBytes); err != nil {
+		fatal(fmt.Errorf("unable to decode macaroon: %v", err))
+	}
+	return mac, nil
+}
+
+// storeMacaroon stores a native macaroon instance to the entry. This will ask
+// the user for an encryption password on the terminal. If a password is typed,
+// the macaroon is stored encrypted with that password.
+func (e *macaroonEntry) storeMacaroon(mac *macaroon.Macaroon,
+	skipEncryption bool) error {
+
+	// First of all, make sure we can serialize the macaroon. 
+	macBytes, err := mac.MarshalBinary()
+	if err != nil {
+		return fmt.Errorf("unable to marshal macaroon: %v", err)
+	}
+
+	if skipEncryption {
+		e.Data = hex.EncodeToString(macBytes)
+		return nil
+	}
+
+	// Read a password from the termina. If it's empty, we won't encrypt the
+	// macaroon and store it plaintext.
+	pw, err := capturePassword(
+		"Enter password to encrypt macaroon with or leave blank to "+
+			"store in plaintext: ", true,
+		walletunlocker.ValidatePassword,
+	)
+	if err != nil {
+		return fmt.Errorf("unable to get encryption password: %v", err)
+	}
+	if pw == nil {
+		e.Data = hex.EncodeToString(macBytes)
+		return nil
+	}
+
+	// The user did set a password. Let's derive an encryption key from it.
+	key, err := snacl.NewSecretKey(
+		&pw, snacl.DefaultN, snacl.DefaultR, snacl.DefaultP,
+	)
+	if err != nil {
+		return fmt.Errorf("unable to create encryption key: %v", err)
+	}
+
+	// Encrypt the macaroon data with the derived key and store it in the
+	// human readable format snacl:<key_base64>:<encrypted_macaroon_base64>.
+	encryptedMac, err := key.Encrypt(macBytes)
+	if err != nil {
+		return fmt.Errorf("unable to encrypt macaroon: %v", err)
+	}
+
+	keyB64 := base64.StdEncoding.EncodeToString(key.Marshal())
+	dataB64 := base64.StdEncoding.EncodeToString(encryptedMac)
+	e.Data = fmt.Sprintf("%s%s:%s", encryptionPrefix, keyB64, dataB64)
+
+	return nil
+}
+
+// decryptMacaroon decrypts the cipher text macaroon by using the serialized
+// encryption key. The user will be prompted for the encryption password on the
+// terminal.
+func decryptMacaroon(keyB64, dataB64 string) ([]byte, error) {
+	// Base64 decode both the marshalled encryption key and macaroon data.
+	keyData, err := base64.StdEncoding.DecodeString(keyB64)
+	if err != nil {
+		return nil, fmt.Errorf("could not base64 decode encryption "+
+			"key: %v", err)
+	}
+	encryptedMac, err := base64.StdEncoding.DecodeString(dataB64)
+	if err != nil {
+		return nil, fmt.Errorf("could not base64 decode macaroon "+
+			"data: %v", err)
+	}
+
+	// Unmarshal the encryption key and ask the user for the password.
+	key := &snacl.SecretKey{}
+	err = key.Unmarshal(keyData)
+	if err != nil {
+		return nil, fmt.Errorf("could not unmarshal encryption key: %v",
+			err)
+	}
+	pw, err := readPassword("Enter macaroon encryption password: ")
```

Comment:
similarly: if we want to test these functions then it's better to pass the password